diff --git a/src/libFLAC/windows_unicode_filenames.c b/src/libFLAC/windows_unicode_filenames.c
index 2404e31..09dd12a 100644
--- a/src/libFLAC/windows_unicode_filenames.c
+++ b/src/libFLAC/windows_unicode_filenames.c
@@ -183,6 +183,42 @@ int flac_internal_rename_utf8(const char *oldname, const char *newname)
 	}
 }
 
+HANDLE WINAPI CreateFileW(LPCWSTR lpFileName,
+                          DWORD dwDesiredAccess,
+                          DWORD dwShareMode,
+                          LPSECURITY_ATTRIBUTES lpSecurityAttributes,
+                          DWORD dwCreationDisposition,
+                          DWORD dwFlagsAndAttributes,
+                          HANDLE hTemplateFile)
+{
+    CREATEFILE2_EXTENDED_PARAMETERS createExParams;
+    createExParams.dwSize = sizeof(CREATEFILE2_EXTENDED_PARAMETERS);
+    createExParams.dwFileAttributes = dwFlagsAndAttributes & 0xFFFF;
+    createExParams.dwFileFlags = dwFlagsAndAttributes & 0xFFF00000;
+    createExParams.dwSecurityQosFlags = dwFlagsAndAttributes & 0x000F00000;
+    createExParams.lpSecurityAttributes = lpSecurityAttributes;
+    createExParams.hTemplateFile = hTemplateFile;
+    return CreateFile2(lpFileName, dwDesiredAccess, dwShareMode, dwCreationDisposition, &createExParams);
+}
+
+HANDLE WINAPI CreateFileA(LPCSTR lpFileName,
+                          DWORD dwDesiredAccess,
+                          DWORD dwShareMode,
+                          LPSECURITY_ATTRIBUTES lpSecurityAttributes,
+                          DWORD dwCreationDisposition,
+                          DWORD dwFlagsAndAttributes,
+                          HANDLE hTemplateFile)
+{
+	HANDLE handle = INVALID_HANDLE_VALUE;
+	wchar_t *wname;
+
+	if ((wname = wchar_from_utf8(lpFileName)) != NULL) {
+		handle = CreateFileW(lpFileName, dwDesiredAccess, dwShareMode, lpSecurityAttributes, dwCreationDisposition, dwFlagsAndAttributes, hTemplateFile);
+		free(wname);
+	}
+	return handle;
+}
+
 HANDLE WINAPI flac_internal_CreateFile_utf8(const char *lpFileName, DWORD dwDesiredAccess, DWORD dwShareMode, LPSECURITY_ATTRIBUTES lpSecurityAttributes, DWORD dwCreationDisposition, DWORD dwFlagsAndAttributes, HANDLE hTemplateFile)
 {
 	if (!utf8_filenames) {
-- 
2.13.2.windows.1

